keras/optimizers.py

```diff
@@ -178,15 +177,26 @@ def get_updates(self, loss, params):
             lr *= (1. / (1. + self.decay * K.cast(self.iterations,
                                                   K.dtype(self.decay))))
         # momentum
+        # TODO(johmathe): Add back nesterov.
         shapes = [K.int_shape(p) for p in params]
         moments = [K.zeros(shape) for shape in shapes]
         self.weights = [self.iterations] + moments
         for p, g, m in zip(params, grads, moments):
-            v = self.momentum * m - lr * g  # velocity
+            v = self.momentum * m - lr * g
             self.updates.append(K.update(m, v))
 
-            if self.nesterov:
-                new_p = p + self.momentum * v - lr * g
+            # Do the gradient descent on the manifold if present.
+            if getattr(p, 'manifold', None) is not None:
+                print('MANIF')
+                shape = K.shape(v)
+                manifold = p.manifold
+                v_shaped = K.reshape(v, (manifold.dimension, -1))
+                p_shaped = K.reshape(p, (manifold.dimension, -1))
+                tangent_v = manifold.projection_to_tangent_space(
```

Comment:
only the manifolds deriving from the class `EmbeddedManifold` have the method `projection_to_tangent_space`.